### PR TITLE
Addressed timeout timing bug with mqtt ping timeouts.

### PR DESCRIPTION
**Pull request by JonathanHenson**
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: source/client.c [hunk 1]
```diff
@@ -851,6 +851,19 @@ int aws_mqtt_client_connection_connect(
             (uint64_t)connection_options->ping_timeout_ms, AWS_TIMESTAMP_MILLIS, AWS_TIMESTAMP_NANOS, NULL);
     }
 
+    /* Keep alive time should always be greater than the timeouts. */
+    if (AWS_UNLIKELY(
```
Comment (justinboswell) [inline source/client.c]: Couldn't this be a warning, and we just clamp the value for you?

File: source/client.c [hunk 2]
```diff
@@ -851,6 +851,19 @@ int aws_mqtt_client_connection_connect(
...
+            connection->keep_alive_time_secs * (uint64_t)AWS_TIMESTAMP_NANOS <= connection->request_timeout_ns)) {
+        AWS_LOGF_FATAL(
+            AWS_LS_MQTT_CLIENT,
+            "id=%p: Illegal configuration, Connection keep alive %" PRIu16
+            "s must be greater than the request timeouts %" PRIu64 "ns.",
+            (void *)connection,
+            connection->keep_alive_time_secs,
+            connection->request_timeout_ns);
+        AWS_FATAL_ASSERT(0);
```
Comment (justinboswell) [inline source/client.c]: If we keep this, just assert the condition above, so that the assert is helpful.

File: source/client.c [hunk 3]
```diff
@@ -1735,22 +1748,17 @@ static enum aws_mqtt_client_request_state s_pingreq_send(uint16_t message_id, bo
...
     }
 
     /* Check that a pingresp has been received since pingreq was sent */
-
-    uint64_t current_time = 0;
-    aws_channel_current_clock_time(connection->slot->channel, &current_time);
-
-    if (current_time - connection->last_pingresp_timestamp > connection->request_timeout_ns) {
+    if (connection->waiting_on_ping_response) {
+        connection->waiting_on_ping_response = false;
```
Comment (justinboswell) [inline source/client.c]: What if multiple pings are in flight? Should this maybe tick up when we send, and down when we receive?

File: include/aws/mqtt/client.h [hunk 4]
```diff
@@ -153,12 +153,11 @@ struct aws_mqtt_topic_subscription {
...
  * keep_alive_time_secs      The keep alive value to place in the CONNECT PACKET, a PING will automatically
  *                           be sent at this interval as well. If you specify 0, defaults will be used
- *                           and a ping will be sent once per 60 minutes.
- * ping_timeout_ms           Network connection is re-established if a ping response is not received within
- *                           this amount of time (milliseconds). If you specify 0, a default value of 3 seconds is used.
- *                           Alternatively, tcp keep-alive may be away to accomplish this in a more efficient
- *                           (low-power) scenario, but keep-alive options may not work the same way on every platform
- *                           and OS version.
+ *                           and a ping will be sent once per 60 minutes. This value must be greater than
+ * ping_timeout_ms ping_timeout_ms           Network connection is re-established if a ping response is not received
```
Comment (bretambrose) [inline include/aws/mqtt/client.h]: Seems like some formatting got mangled here

File: include/aws/mqtt/client.h [hunk 5]
```diff
@@ -153,12 +153,11 @@ struct aws_mqtt_topic_subscription {
...
+ * within this amount of time (milliseconds). If you specify 0, a default value of 3 seconds is used. Alternatively, tcp
+ * keep-alive may be away to accomplish this in a more efficient (low-power) scenario, but keep-alive options may not
```
Comment (bretambrose) [inline include/aws/mqtt/client.h]: might as well correct "away" -> "a way" while here

File: source/client.c [hunk 6]
```diff
@@ -851,6 +851,19 @@ int aws_mqtt_client_connection_connect(
             (uint64_t)connection_options->ping_timeout_ms, AWS_TIMESTAMP_MILLIS, AWS_TIMESTAMP_NANOS, NULL);
     }
 
+    /* Keep alive time should always be greater than the timeouts. */
+    if (AWS_UNLIKELY(
+            connection->keep_alive_time_secs * (uint64_t)AWS_TIMESTAMP_NANOS <= connection->request_timeout_ns)) {
+        AWS_LOGF_FATAL(
+            AWS_LS_MQTT_CLIENT,
+            "id=%p: Illegal configuration, Connection keep alive %" PRIu16
+            "s must be greater than the request timeouts %" PRIu64 "ns.",
```
Comment (bretambrose) [inline source/client.c]: It's annoying, but I'd make sure both time values were in the same unit here, to make interpretation easier and less likely to be wrong.

File: source/client.c (hunk 3 above)
Comment (JonathanHenson) [reply source/client.c]: not possible now since PING is being removed from the public API

File: source/client.c (hunk 1 above)
Comment (JonathanHenson) [reply source/client.c]: That's going to be a hidden behavior that we can't actually get right 100% of the time. We already have default values, if someone sets them, and they're wrong, I think we should let them know immediately.

Comment (justinboswell) [reply source/client.c]: Ok, fair enough, sounds good.

File: include/aws/mqtt/private/client_impl.h [hunk 7]
```diff
@@ -201,4 +201,16 @@ AWS_MQTT_API void mqtt_request_complete(
...
+ * Sends a PINGREQ packet to the server to keep the connection alive. This is not exported and should not ever
+ * be called directly. This function is driven by the timeout values passed to aws_mqtt_client_connect().
+ * If a PINGRESP is not received within a reasonable period of time, the connection will be closed.
+ *
+ * \params[in] connection   The connection to ping on
+ *
+ * \returns AWS_OP_SUCCESS if the connection is open and the PINGREQ is sent or queued to send,
+ *              otherwise AWS_OP_ERR and aws_last_error() is set.
+ */
+int aws_mqtt_client_connection_ping(struct aws_mqtt_client_connection *connection);
```
Comment (justinboswell) [inline include/aws/mqtt/private/client_impl.h]: Remove the aws_ since it's private now?

**Pull request merged**